ts/protocol/utils.ts

```diff
@@ -192,3 +192,20 @@ export function timeoutPromise<T>(error: Error, timeoutMs: number): Promise<T> {
   let promiseResolver = new TimeoutPromiseResolver(error, timeoutMs);
   return promiseResolver.promise;
 }
+
+/**
+ * Is a cancellable PromiseResolver.
+ */
+export class CancellablePromise<T> extends PromiseResolver<T> {
+  constructor(private readonly value: T) {
+    super();
+  }
+
+  cancel() {
+    this.resolve(this.value);
```

Comment:
We do. The TimeoutPromiseResolver rejects on timeout: https://github.com/openid/OpenYOLO-Web/blob/master/ts/protocol/utils.ts#L183 - or where you talking about something different? (we also reject promise very often across the codebase)

When you mention result codes, are you saying *error codes*? The results are determined by the type of the Promise (i.e. `Credential|null`, `boolean`, etc.), and cannot contain a *code*. The errors, however, would benefit from having codes.